pkg/k8sutil/rollout.go

```diff
@@ -0,0 +1,223 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	clientappsv1 "k8s.io/client-go/kubernetes/typed/apps/v1"
+	"k8s.io/client-go/tools/cache"
+	watchtools "k8s.io/client-go/tools/watch"
+)
+
+// RolloutRestartDaemonSet is the programmatic equivalent of 'kubectl rollout restart'.
+func RolloutRestartDaemonSet(dsi clientappsv1.DaemonSetInterface, name string) (int64, error) {
+	ds, err := dsi.Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return 0, fmt.Errorf("getting DaemonSet: %w", err)
+	}
+
+	if ds.Spec.Template.Annotations == nil {
+		ds.Spec.Template.Annotations = map[string]string{}
+	}
+
+	ds.Spec.Template.Annotations["kubectl.kubernetes.io/restartedAt"] = time.Now().String()
+
+	newDS, err := dsi.Update(context.TODO(), ds, metav1.UpdateOptions{})
+	if err != nil {
+		return 0, fmt.Errorf("updating DaemonSet: %w", err)
+	}
+
+	return newDS.Generation, nil
+}
+
+// RolloutRestartDeployment is the programmatic equivalent of 'kubectl rollout restart'.
+func RolloutRestartDeployment(deployClient clientappsv1.DeploymentInterface, name string) (int64, error) {
+	d, err := deployClient.Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return 0, fmt.Errorf("getting Deployment: %w", err)
+	}
+
+	if d.Spec.Template.Annotations == nil {
+		d.Spec.Template.Annotations = map[string]string{}
+	}
+
+	d.Spec.Template.Annotations["kubectl.kubernetes.io/restartedAt"] = time.Now().String()
+
+	newDeploy, err := deployClient.Update(context.TODO(), d, metav1.UpdateOptions{})
+	if err != nil {
+		return 0, fmt.Errorf("updating Deployment: %w", err)
+	}
+
+	return newDeploy.Generation, nil
+}
+
+// daemonSetUpToDate checks if given DaemonSet converged.
+func daemonSetUpToDate(ds *appsv1.DaemonSet, generation int64) (bool, error) {
+	if ds.Spec.UpdateStrategy.Type != appsv1.RollingUpdateDaemonSetStrategyType {
+		return true, fmt.Errorf("rollout status is only available for %s strategy type",
+			appsv1.RollingUpdateStatefulSetStrategyType)
+	}
+
+	if generation > 0 && ds.Status.ObservedGeneration < generation {
+		return false, nil
+	}
+
+	replicas := ds.Status.DesiredNumberScheduled
+
+	if replicas == 0 {
+		return false, nil
+	}
+
+	if ds.Status.NumberReady == replicas && ds.Status.UpdatedNumberScheduled == replicas {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// WaitOptions holds optional arguments for WaitFor... functions group.
+type WaitOptions struct {
+	Generation int64
+	Timeout    time.Duration
+}
+
+const (
+	// WaitDefaultTimeout is default timeout after which watching for workload should return error.
+	WaitDefaultTimeout = 5 * time.Minute
+)
+
+// WaitForDaemonSet waits until DaemonSet converges.
+//
+//nolint:dupl
+func WaitForDaemonSet(dsi clientappsv1.DaemonSetInterface, name string, options WaitOptions) error {
```

Comment:
I think we should have e2e tests for this functions, including restarting the controller manager and scheduler, so we can verify it behaves correctly when schedulers are disrupted, as this is basically what we do when we rotate the certificates.